test/functionalTest/cases/3190_notify_only_attributes_that_change/notify_only_attributes_that_change_erros.test

```diff
@@ -0,0 +1,187 @@
+# Copyright 2019 Telefonica Investigacion y Desarrollo, S.A.U
+#
+# This file is part of Orion Context Broker.
+#
+# Orion Context Broker is free software: you can redistribute it and/or
+# modify it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# Orion Context Broker is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero
+# General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+#
+# For those usages not covered by this license please contact with
+# iot_support at tid dot es
+
+# VALGRIND_READY - to mark the test ready for valgrindTestSuite.sh
+
+--NAME--
+Notify only attributes that change for NGSIv2 for error POST /v2/subscriptions, POST /v2/subscriptions, POST /v2/subscriptions.
+
+--SHELL-INIT--
+dbInit CB
+brokerStart CB
+accumulatorStart --pretty-print localhost $LISTENER_PORT
+
+--SHELL--
+
+#
+# 01. POST /v2/subscriptions, create subscription with attrs, exceptAttrs and onlyChangedAttrs.
+# 02. POST /v2/subscriptions, create subscription with onlyChangedAttrs.
+# 03. POST /v2/subscriptions, create subscription with attrs and onlyChangedAttrs set to incorrect value.
+#
+
+echo '01. POST /v2/subscriptions, create subscription with attrs, exceptAttrs and onlyChangedAttrs.'
+echo '============================================================================================='
+payload='{
+  "description": "A subscription to get info about Room1",
+  "subject": {
+    "entities": [
+      {
+        "id": "Room1",
+        "type": "Room"
+      }
+    ],
+    "condition": {
+      "attrs": [
+        "A","B","C"
+      ]
+    }
+  },
+  "notification": {
+    "http": {
+      "url": "http://localhost:'$LISTENER_PORT'/notify"
+    },
+	"attrs" : [
+	  "A","B"
+	],
+	"exceptAttrs" : [
+	  "C"
+	],
+    "onlyChangedAttrs":true
+  },
+  "expires": "2040-01-01T14:00:00.00Z",
+  "throttling": 5
+}'
+orionCurl --url /v2/subscriptions --payload "$payload"
+echo
+echo
+
+
+echo '02. POST /v2/subscriptions, create subscription with onlyChangedAttrs.'
+echo '======================================================================'
+payload='{
+  "description": "A subscription to get info about Room1",
+  "subject": {
+    "entities": [
+      {
+        "id": "Room1",
+        "type": "Room"
+      }
+    ],
+    "condition": {
+      "attrs": [
+        "A","B","C"
+      ]
+    }
+  },
+  "notification": {
+    "http": {
+      "url": "http://localhost:'$LISTENER_PORT'/notify"
+    },
+    "onlyChangedAttrs":true
+  },
+  "expires": "2040-01-01T14:00:00.00Z",
+  "throttling": 5
+}'
+orionCurl --url /v2/subscriptions --payload "$payload"
```

Comment:
This situation is legal. This means that Orion will potentially include in the notification any attribute, but only those that change.

In fact, I'd suggest the following use cases (maybe in a separate .test).

* Entity has attributes A, B, C and D
* Subscription is cond.attrs: [A, B, C], notification.attrs: not included, ondlyChanged: true
* Update changing D -> no notification is sent (because D is not in cond.attrs)
* Update changing A, B -> notification including A and B
* Update changing A, D -> notification including A and D (note that although D is not part of cond.attrs, is an attribute that has changed and there is no notificadtion.attrs in place to filter this)

What do you think? Does it make sense?

Reply:
Hi @fgalan Sir,

As suggested by you earlier on [issue](https://github.com/telefonicaid/fiware-orion/issues/3190#issuecomment-490433026), I have implemented the same.

```
The updates changes A and B then only B will be notified (as it is the one 
that match the conditions "has changed" and "is in notification.attrs filter" 
at the same time).
```
According to this, our understanding is that we have to filter the common attributes from modified-attributes[A,B], condition.attrs[A,B,C] and notification.attrs[B,C,D] and notify them (in this case notification will be sent to B).

But according to the below comment,
```
Update changing A, D -> notification including A and D (note that although 
D is not part of cond.attrs, is an attribute that has changed and there are no 
notification.attrs in place to filter this)
```
A and D should be notified. 

But as per [document](https://fiware-orion.readthedocs.io/en/master/user/walkthrough_apiv2/index.html#subscriptions)

```
notifications.attrs list (or to use an empty attribute vector, which means 
"all the attributes in the entity")
```

Since the notification.attrs[] is empty. it means that we should consider all the attributes[A,B,C,D] belonging to that entity.

So as suggested by you earlier we should filter the common attributes from modified-attributes[A,D], condition.attrs[A,B,C] and notification.attrs[A,B,C,D] and notify only A.  

Could you please confirm the above understanding and guide me further on this?

Thanks

Reply:
I think [the case you cite in the issue](https://github.com/telefonicaid/fiware-orion/issues/3190#issuecomment-490433026) doesn't applies here, as that one has notification.attrs not empty and the case I'm describing [above](https://github.com/telefonicaid/fiware-orion/pull/3490#discussion_r286967087) is about notification.attrs empty.

I have to admit this is a bit complicated to describe :)  but maybe it could be easily understood if we model it using "filter", from a conceptual point of view. Let's consider the following filter:

* filter_notif(A1, A2..., An) -> this filter returns attributes in A1, A2,... An that are in the notifications.attr list. If notification.attrs is empty, it 
* filter_changes(A1, A2, ..., An) -> this filter returns attributes in A1, A2,... An that has changed

-----

Now, let's consider and entity with A, B, C, D attributes and the case of a conventional subscription (i.e. one without onlyChanged: true) with condition.attrs: [A, B, C] and notifications.attrs = [ ]. An update in A and D arrives. That's triggers the subsciprtion, as A is in conditions.attrs list.

The attributes to notify will be the ones returned by the notifications filter (i.e. all attributes of the entity):

filter_notif(A, B, C, D) = [A, B, C, D]

-----

Now, same situation with the subscription with onlyChanged: true (the one I describe [above](https://github.com/telefonicaid/fiware-orion/pull/3490#discussion_r286967087)). Now we have to combine both:

filter_notif(filter_change(A, B, C, D)) = filter_notif(A, D) = [A, D]

or the other way arround:

filter_change(filter_notif(A, B, C, D)) = filter_change(A, B, C, D) = [A, B]

----

Now let's consider the [the case you cite in the issue](https://github.com/telefonicaid/fiware-orion/issues/3190#issuecomment-490433026). In thic ase notifications.attr = [B, C, D]. Let's consider update only in A and B. In this case:

filter_notif(filter_change(A, B, C, D)) = filter_notif(A, B) = [B]

or

filter_change(filter_notif(A, B, C, D)) = filter_change(B, C, D) = [B]

----

Does it make sense? If you find some flaw in my reasoning (I'm not always right! :) please tell me.

(I'm not meaning to implement these filters in code. What I mean is that from a conceptual point of view test cases should behave like that).

Reply:
fixed in f9887f1